File: ws04_tools/src/tf_listener/src/tf_listener_node.cpp
#include <memory>
#include <string>
#include <chrono>
#include <iostream>

#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/exceptions.h"

using namespace std::chrono_literals;

class TfListenerNode : public rclcpp::Node
{
public:
  TfListenerNode()
  : Node("tf_listener")
  {
    buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
    listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_);

    // 参数：父帧、子帧（可通过参数修改）
    this->declare_parameter<std::string>("parent_frame", "world");
    this->declare_parameter<std::string>("child_frame", "child_frame");
    parent_frame_ = this->get_parameter("parent_frame").as_string();
    child_frame_ = this->get_parameter("child_frame").as_string();

    timer_ = this->create_wall_timer(500ms, std::bind(&TfListenerNode::on_timer, this));

    RCLCPP_INFO(this->get_logger(), "tf_listener started, parent: '%s', child: '%s'", parent_frame_.c_str(), child_frame_.c_str());
  }

private:
  void on_timer()
  {
    geometry_msgs::msg::TransformStamped transform;
    try {
      // 使用最新的可用变换
      transform = buffer_->lookupTransform(parent_frame_, child_frame_, tf2::TimePointZero);

      auto &t = transform.transform.translation;
      auto &r = transform.transform.rotation;

      RCLCPP_INFO(this->get_logger(), "Transform %s -> %s: translation=(%.3f, %.3f, %.3f) rotation=(%.3f, %.3f, %.3f, %.3f)",
        transform.header.frame_id.c_str(), transform.child_frame_id.c_str(),
        t.x, t.y, t.z, r.x, r.y, r.z, r.w);
    } catch (const tf2::TransformException &ex) {
      RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000, "Could not get transform: %s", ex.what());
    }
  }

  std::shared_ptr<tf2_ros::Buffer> buffer_;
  std::shared_ptr<tf2_ros::TransformListener> listener_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::string parent_frame_;
  std::string child_frame_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<TfListenerNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: ws04_tools/src/tf_listener_py/tf_listener_py/tf_listener_node.py
import rclpy
from rclpy.node import Node
from tf2_ros import Buffer, TransformListener
from tf2_ros import TransformException
import time

class TFListenerPy(Node):
    def __init__(self):
        super().__init__("tf_listener_py_node")
        # 创建buffer
        self.buffer = Buffer()
        # 创建transform_listener
        self.listener = TransformListener(self.buffer, self)
        
        # 参数：父帧、子帧（可通过参数修改）
        self.declare_parameter("parent_frame", "world")
        self.declare_parameter("child_frame", "child_frame")
        self.parent_frame = self.get_parameter("parent_frame").get_parameter_value().string_value
        self.child_frame = self.get_parameter("child_frame").get_parameter_value().string_value

        # 创建定时器
        self.timer = self.create_timer(0.5, self.on_timer)

        self.get_logger().info(f"tf_listener started, parent: '{self.parent_frame}', child: '{self.child_frame}'")

    def on_timer(self):
        try:
            # 使用最新的可用变换
            transform = self.buffer.lookup_transform(self.parent_frame, self.child_frame, rclpy.time.Time())
            
            t = transform.transform.translation
            r = transform.transform.rotation
            
            self.get_logger().info(
                f"Transform {transform.header.frame_id} -> {transform.child_frame_id}: "
                f"translation=({t.x:.3f}, {t.y:.3f}, {t.z:.3f}) "
                f"rotation=({r.x:.3f}, {r.y:.3f}, {r.z:.3f}, {r.w:.3f})"
            )
        except TransformException as ex:
            self.get_logger().warn(f"Could not get transform: {ex}", throttle_duration_sec=2)


def main(args=None):
    rclpy.init(args=args)
    node = TFListenerPy()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()